src/container.cpp

```diff
@@ -443,9 +443,16 @@ Cylinder* Container::queryDestination(int32_t& index, const Thing &thing, Item**
 		return this;
 	}
 
+	if (index != INDEX_WHEREEVER) {
+		Item* itemFromIndex = getItemByIndex(index);
+		if (itemFromIndex) {
+			*destItem = itemFromIndex;
+		}
+	}
+
 	bool autoStack = !hasBitSet(FLAG_IGNOREAUTOSTACK, flags);
-	if (autoStack && item->isStackable() && item->getParent() != this) {
-		//try find a suitable item to stack with
+	//try find a suitable item to stack with
+	if (autoStack && item->isStackable() && item->getParent() != this && !(*destItem && (*destItem)->isStackable() && (*destItem)->equals(item))) {
```

Comment:
!(*destItem && (*destItem)->isStackable() && (*destItem)->equals(item))
Can't this be simplified to:
!(*destItem && (*destItem)->equals(item))

I can't verify this right now, but afaik, ->equals should be enough for this check